src/universe.rs

```diff
@@ -94,24 +108,96 @@ impl SymbolStore {
         self.name_by_sym.push(None);
         sym
     }
+}
 
+impl Symbols for SymbolStore {
     /// Get the name of a symbol, if it has one.
-    pub fn get_symbol_name(&self, sym: Sym) -> Option<&str> {
+    fn get_symbol_name(&self, sym: Sym) -> Option<&str> {
         self.name_by_sym.get(sym.ord()).and_then(|n| n.as_deref())
     }
 
     /// Returns the number of symbols that have been allocated in this universe.
-    pub fn num_symbols(&self) -> usize {
+    fn num_symbols(&self) -> usize {
         self.name_by_sym.len()
     }
 }
 
+impl<'a> SymbolStorage for SymbolStore {
+    /// Return the symbol associated with the name, or allocate a fresh ID and associate it with the
+    /// given name.
+    fn get_or_insert_named(&mut self, name: &str) -> Sym {
+        self.sym_by_name.get(name).copied().unwrap_or_else(|| {
+            let sym = Sym::from_ord(self.name_by_sym.len());
+            let shared_name: Arc<str> = name.into();
+            self.name_by_sym.push(Some(shared_name.clone()));
+            self.sym_by_name.insert(shared_name, sym);
+            sym
+        })
+    }
+}
+
 impl Default for SymbolStore {
     fn default() -> Self {
         Self::new()
     }
 }
 
+/// Stores and allocates unique symbols on top of a symbol store, without modifying it.
+#[derive(Debug)]
+pub struct SymbolOverlay<'a> {
+    symbols: &'a SymbolStore,
+    overlay: SymbolStore,
+}
+
+impl<'a> SymbolOverlay<'a> {
+    pub fn new(symbols: &'a SymbolStore) -> Self {
+        Self {
+            symbols,
+            overlay: Default::default(),
+        }
+    }
+}
+
+impl SymbolStorage for SymbolOverlay<'_> {
+    /// Return the symbol associated with the name, or allocate a fresh ID and associate it with the
+    /// given name.
+    fn get_or_insert_named(&mut self, name: &str) -> Sym {
+        self.symbols
+            .sym_by_name
+            .get(name)
+            .copied()
+            .unwrap_or_else(|| {
+                self.overlay
+                    .sym_by_name
+                    .get(name)
+                    .copied()
+                    .unwrap_or_else(|| {
+                        let sym = Sym::from_ord(self.num_symbols());
+                        let shared_name: Arc<str> = name.into();
+                        self.overlay.name_by_sym.push(Some(shared_name.clone()));
+                        self.overlay.sym_by_name.insert(shared_name, sym);
+                        sym
+                    })
+            })
+    }
+}
+
+impl Symbols for SymbolOverlay<'_> {
+    /// Get the name of a symbol, if it has one.
+    fn get_symbol_name(&self, sym: Sym) -> Option<&str> {
+        self.symbols.get_symbol_name(sym).or_else(||
+            // can't just call self.overlay.get_symbol_name because the array indices are now offset
+            sym.ord().checked_sub(self.symbols.num_symbols())
+                .and_then(|index| self.overlay.name_by_sym.get(index))
+                .and_then(|n| n.as_deref()))
```

Comment:
Speaking of `get_symbol_name`, I think the order of operations here is "wrong" (in the sense of "unclear regarding the intended semantics", not in terms of the result):
The `checked_sub` should come first, and depending on the outcome, it should either look in `symbols` or in `overlay`.

The way it currently works, we still do the `checked_sub` even if `symbols` had the symbol, but just no name for it. But in that case, the `checked_sub` would return `None` anyway, so it's redundant.

Reply:
Done.